src/components/route-book/index.tsx

```diff
@@ -0,0 +1,175 @@
+        amount: Math.min(entry.amount, route.amount),
+      });
+    } else {
+      totals.set(route.price, {
+        ...route,
+        total: route.amount,
+      });
+    }
+  });
+
+  return isBuySide
+    ? Array.from(totals.values())
+        .slice(0, limit)
+        .toSorted((a, b) => b.price - a.price)
+    : Array.from(totals.values()).slice(-limit);
+}
+
+function getDisplayData({
+  data,
+  computePositionId,
+  asset1,
+  asset2,
+  isBuySide,
+  limit,
+}: {
+  data: Position[];
+  computePositionId: ((position: Position) => PositionId) | undefined;
+  asset1: Metadata | undefined;
+  asset2: Metadata | undefined;
+  isBuySide: boolean;
+  limit: number;
+}): RouteWithTotal[] {
+  if (!computePositionId || !asset1 || !asset2) {
+    return [];
+  }
+
+  const asset1Exponent = getDisplayDenomExponent(asset1);
+  const asset2Exponent = getDisplayDenomExponent(asset2);
+
+  const getValue = (property: Amount | undefined, exponent: number) =>
+    fromBaseUnit(BigInt(property?.lo ?? 0), BigInt(property?.hi ?? 0), exponent);
+
+  const routes = data
+    // .filter(position => position.state?.state.toLocaleString() === 'POSITION_STATE_ENUM_OPENED')
+    .filter(position => position.state?.state === PositionState_PositionStateEnum.OPENED)
+    .map(position => {
+      const direction = position.phi?.pair?.asset2?.equals(asset1.penumbraAssetId) ? -1 : 1;
+
+      const r1 = getValue(position.reserves?.r1, direction === 1 ? asset1Exponent : asset2Exponent);
+      const r2 = getValue(position.reserves?.r2, direction === 1 ? asset2Exponent : asset1Exponent);
+      const p = getValue(
+        position.phi?.component?.p,
+        direction === 1 ? asset2Exponent : asset1Exponent,
+      );
+      const q = getValue(
+        position.phi?.component?.q,
+        direction === 1 ? asset1Exponent : asset2Exponent,
+      );
+      const price = Number(direction === 1 ? p.div(q) : q.div(p));
+      const amount = isBuySide
+        ? Number(direction === 1 ? r2 : r1) / price
+        : Number(direction === 1 ? r1 : r2);
+
+      const id = computePositionId(position);
+      const innerStr = uint8ArrayToBase64(id.inner);
+      const bech32Id = innerToBech32Address(innerStr, 'plpid');
+
+      return {
+        lpId: bech32Id,
+        position,
+        price,
+        amount,
+      };
+    })
+    .filter(displayData => displayData.amount > 0)
+    .toSorted((a, b) => b.price - a.price) as Route[];
+
+  return getTotals(routes, isBuySide, limit);
+}
+
+export function RouteBook() {
+  const { data: assets } = useAssets();
+  const asset1 = assets?.find(asset => asset.symbol === 'UM');
+  const asset2 = assets?.find(asset => asset.symbol === 'GM');
+  const asset1Exponent = asset1 ? getDisplayDenomExponent(asset1) : 0;
+  const asset2Exponent = asset2 ? getDisplayDenomExponent(asset2) : 0;
+  const { data: computePositionId } = useComputePositionId();
+  const { data } = useBook(asset1?.symbol, asset2?.symbol, 100, 50);
```

Comment:
suggestion: related to my other comment, but I think we should consider:

```ts
if (!asset1?.symbol || !asset2?.symbol) {
   return <div>No asset selected</div>
}
```

Note that this comes before `useBook()`, and given the rule of hooks, you'd have to create another component after the if clause.

Reply:
I see your point, if you don't mind I will address this in the next update (separate pr).